File: 0008/StringToInteger_atoi.swift
class Solution {
    func myAtoi(_ str: String) -> Int {
        var number = 0
        let sign: Int

        var iterator = str.makeIterator()
        var currentChar = iterator.next()

        // Discard leading whitespaces
        while let char = currentChar, char.isWhitespace { currentChar = iterator.next() }

        // Check sign
        sign = currentChar == "-" ? -1 : 1
        if currentChar == "+" || currentChar == "-" { currentChar = iterator.next() }

        // Convert digits into an integer
        while let digit = currentChar?.wholeNumberValue {
            number = number * 10 + digit
            guard number <= Int32.max else {
                return Int(sign < 0 ? Int32.min : Int32.max)
            }
            currentChar = iterator.next()
        }

        return sign * number
    }
}

File: 0008/StringToInteger_atoi.c
#include <ctype.h>
#include <limits.h>

int myAtoi(char * s) {
    long int number = 0;

    // Discard leading whitespaces
    while (isspace(*s)) ++s;

    // Check sign
    int sign = 1;
    if (*s == '-' || *s == '+') {
        sign = (*s == '-') ? -1 : 1;
        ++s;
    }

    // Convert digits into an integer
    while (isdigit(*s)) {
        int digit = *s - '0';
        number = number * 10 + digit;
        if (number > INT_MAX) {
            return (sign < 0) ? INT_MIN : INT_MAX;
        }
        ++s;
    }

    return sign * number;
}
